### DISCUSSION TITLE
Failed to connect to server

### DISCUSSION THREAD

**Discussion by MarcelRobitaille** [chipsenkbeil/distant#236] [Q&A] [answered]
This plugin is so promising. It would totally change my life. I'm experiencing enormous lag sshing into my virtual machine and running nvim like that. I really hope this plugin can solve this for me. But enough about that.

I have been watching your videos and reading the docs for a few hours. My client is Windows and I'm trying to connect to a Debian VM running on the same machine (eventually I'd like to connect to a docker container inside the VM, but baby steps). I have the nvim plugin installed, the executable installed on the client and the server, and the system service set up.

I am trying to do `distant launch ssh://user@127.0.0.1:1234`, I get this:
```
Password for user@127.0.0.1:
Failed to connect to server

Cause by:
    Es konnte keine Verbindung hergestellt werden, da der Zielcomputer die Verbindung verweigerte. (os error 10061)
```

When I try from nvim, I get:
```
Exited unexpectedly: exit code 1
```

The ssh port is 1234. I opened port 8080 to the VM. I tried running `distant server listen`, but that didn't seem to help.

I put `--log-level trace --log-file ./log.txt` but there isn't that much more information. How can I troubleshoot this?

```
[2024-01-26 23:45:43.936870 +01:00] TRACE [distant-net\src\client.rs:301] Client receiving (id:13666876371593364180 | origin: 14854035303080068083): ��type�launched�destination�[distant://:788957b2112c573414e3d082f8aacfa4d1c244cc1e34b54d50a17d726e42cec6@127.0.0.1:37203
[2024-01-26 23:45:43.936939 +01:00] TRACE [distant-net\src\client.rs:324] Client delivered response 13666876371593364180 to 14854035303080068083
[2024-01-26 23:45:43.936981 +01:00] TRACE [src\cli\commands\client.rs:151] Updating host to "127.0.0.1" from non-global "127.0.0.1"
[2024-01-26 23:45:43.936995 +01:00] DEBUG [src\cli\commands\client.rs:165] Connecting to server at distant://:788957b2112c573414e3d082f8aacfa4d1c244cc1e34b54d50a17d726e42cec6@127.0.0.1:37203
[2024-01-26 23:45:43.937000 +01:00] TRACE [distant-net\src\manager\client.rs:133] connect(distant://:788957b2112c573414e3d082f8aacfa4d1c244cc1e34b54d50a17d726e42cec6@127.0.0.1:37203, )
[2024-01-26 23:45:43.939712 +01:00] TRACE [distant-net\src\client.rs:357] Client sending ��id�17279019336323331442�payload��type�connect�destination�[distant://:788957b2112c573414e3d082f8aacfa4d1c244cc1e34b54d50a17d726e42cec6@127.0.0.1:37203�options�
[2024-01-26 23:45:45.956929 +01:00] TRACE [distant-net\src\client.rs:301] Client receiving (id:871356558162164671 | origin: 17279019336323331442): ��type�error�description�oEs konnte keine Verbindung hergestellt werden, da der Zielcomputer die Verbindung verweigerte. (os error 10061)
[2024-01-26 23:45:45.956971 +01:00] TRACE [distant-net\src\client.rs:324] Client delivered response 871356558162164671 to 17279019336323331442
[2024-01-26 23:45:45.958644 +01:00] ERROR [src\lib.rs:143] Failed to connect to server

Caused by:
    Es konnte keine Verbindung hergestellt werden, da der Zielcomputer die Verbindung verweigerte. (os error 10061)
```

Also, I have an ssh key set up, but it asks me for my password every time. Is that normal?

**Comment by chipsenkbeil** [comment, marked as answer]
The ssh client code we're using (written in C) has been a bit unreliable, so there's [work to update it](https://github.com/chipsenkbeil/distant/issues/193) (found in the [RusshSupport branch](https://github.com/chipsenkbeil/distant/tree/feat/RusshSupport)). In addition, since you're interested in Docker, there's also work to [support Docker directly](https://github.com/chipsenkbeil/distant/issues/194) instead of needing to host the distant server within containers or use ssh.

In the meantime, the error you're getting appears to be refusal to connect to the TCP port, which means that the server may not be listening properly. Here's what I would try:

1. Make sure that a distant manager is running on your host machine. For now, you can run this in the foreground while we test:
    ```bash
    distant manager listen
    ```
2. Open a port on your VM like you did. In this case, 8080.
3. Run the server within the VM, but run it with several arguments to listen on ANY interface, and explicitly on port 8080:
    ```bash
    distant server listen --port 8080 --host any

    # Should output something like below. Notice 0.0.0.0 as the IP address and 8080 as the port
    # distant://:1c7d6b7435cc86aba80afb9d7a67fb27e7ad6a3a800f3904511fadff03f43022@0.0.0.0:8080
    ```
4. From outside the VM, try to connect to it using distant via whatever port you mapped 8080 to within the VM. For example, if I mapped 12345 on my host machine to 8080 on my VM, then I would connect via
    ```bash
    distant connect distant://:1c7d6b7435cc86aba80afb9d7a67fb27e7ad6a3a800f3904511fadff03f43022@127.0.0.1:12345

    # Should see something like this output, which is the unique id that the manager knows maps to your server
    2449591014

    # Alternative, if your VM is on some other IP address on your local network, say 192.168.1.123, then you need to specify that
    distant connect distant://:1c7d6b7435cc86aba80afb9d7a67fb27e7ad6a3a800f3904511fadff03f43022@192.168.1.123:12345
    ```
5. Finally, test it by doing something!
    ```bash
    # Read the current directory of the server
    distant fs read .
    ```

**Reply by MarcelRobitaille** [reply]
Thank you very much for the answer. Will the docker support still work if the docker container is running in a VM? I guess I just have to forward the right ports, right?

Thank you for the instructions.

Now I'm getting:

```bash
distant server listen --port 8080 --host any
# [flexi_logger][ERRCODE::Write] writing failed, caused by Os { code: 30, kind: ReadOnlyFilesystem, message: "Read-only file system" }
#     See https://docs.rs/flexi_logger/latest/flexi_logger/error_info/index.html#write

# distant://:9c5d10af60c58091442d0d7fa85a46d3d06b148428c89fd51bfe00ae639e2f52@0.0.0.0:8080

# [flexi_logger][ERRCODE::Write] writing failed, caused by Os { code: 30, kind: ReadOnlyFilesystem, message: "Read-only file system" }
#     See https://docs.rs/flexi_logger/latest/flexi_logger/error_info/index.html#write
# [flexi_logger][ERRCODE::Write] writing failed, caused by Os { code: 30, kind: ReadOnlyFilesystem, message: "Read-only file system" }
#     See https://docs.rs/flexi_logger/latest/flexi_logger/error_info/index.html#write
```

And
```bash
distant connect distant://:13c6b7162b82209bfee6e468eb1e568e5304cb3c630abdbe71db0089ee686152@127.0.0.1:8080
# Failed to connect to server

# Caused by:
#     Client version 0.20.0 is incompatible with server version 74.9417417626037023333.8319108863997473913
```

But I think I can fix this one by myself.

**Reply by MarcelRobitaille** [reply]
It is working now! I'm even able to connect with nvim. `DistantSessionInfo` shows that I'm connected. Now I have to learn how to actually do stuff...
